src/aws-codeartifact/repository.ts

```diff
@@ -0,0 +1,430 @@
+  /**
+   * The description of the repository
+   */
+  readonly description?: string;
+  /**
+   * The connections to external repositories (like npmjs, pypi, etc.)
+   *
+   * You can use the AWS CLI to connect your CodeArtifact repository to an external repository by adding an external connection directly to the repository.
+   * This will allow users connected to the CodeArtifact repository, or any of its downstream repositories, to fetch packages from the configured external repository.
+   * Each CodeArtifact repository can only have one external connection.
+   */
+  readonly externalConnection?: RepositoryConnection;
+  /**
+   * A list of upstream Codeartifact repositories to associate with the repository.
+   * The order of the upstream repositories in the list determines their priority order when CodeArtifact looks for a requested package version.
+   * see https://docs.aws.amazon.com/codeartifact/latest/ug/repo-upstream-behavior.html#package-retention-intermediate-repositories
+   *
+   * @default - No upstream repositories
+   */
+  readonly upstreams?: IRepository[];
+}
+
+/**
+ * A new or imported Codeartifact Repository.
+ */
+abstract class RepositoryBase extends Resource implements IRepository {
+  /**
+   * The ARN of the repository
+   */
+  public abstract repositoryArn: string;
+  /**
+   * The name of the repository
+   */
+  public abstract repositoryName: string;
+  /**
+   * The domain that contains the repository
+   */
+  public abstract repositoryDomainName: string;
+  /**
+   * The domain owner of the repository
+   */
+  public abstract repositoryDomainOwner: string;
+  /**
+   * The domain that contains the repository
+   */
+  public abstract domain: IDomain;
+
+  /**
+   * Optional policy document that represents the resource policy of this repository.
+   *
+   * If specified, addToResourcePolicy can be used to edit this policy.
+   * Otherwise this method will no-op.
+   */
+  protected abstract readonly policy?: PolicyDocument;
+
+  private isCrossEnvironmentGrantee(grantee: IGrantable): boolean {
+    if (!principalIsOwnedResource(grantee.grantPrincipal)) {
+      return false;
+    }
+    const thisStack = Stack.of(this);
+    const identityStack = Stack.of(grantee.grantPrincipal);
+    return thisStack.region !== identityStack.region || thisStack.account !== identityStack.account;
+  }
+
+  /**
+   * Adds a statement to the CodeArtifact repository resource policy.
+   * @param statement The policy statement to add
+   */
+  public addToResourcePolicy(statement: PolicyStatement): AddToResourcePolicyResult {
+    const stack = Stack.of(this);
+
+    if (!this.policy) {
+      Annotations.of(stack).addWarningV2(
+        'NoResourcePolicyStatementAdded',
+        `No statements added to imported resource ${this.repositoryArn}.`,
+      );
+      return { statementAdded: false };
+    }
+
+    this.policy.addStatements(statement);
+    return { statementAdded: true, policyDependable: this.policy };
+  }
```

Comment:
How about this?

```suggestion
  public addToResourcePolicy(statement: PolicyStatement): AddToResourcePolicyResult {
    if (!Resource.isOwnedResource(this)) {
      Annotations.of(Stack.of(this)).addWarningV2(
        'NoResourcePolicyStatementAdded',
        `No statements added to imported resource ${this.repositoryArn}.`,
      );
      return { statementAdded: false };
    }

    if (!this.policy) {
      this.policy = new PolicyDocument();
    }
    this.policy.addStatements(statement);
    return { statementAdded: true, policyDependable: this.policy };
  }
```

And

`protected abstract readonly policy?: PolicyDocument;` -> `protected abstract policy?: PolicyDocument;`